### PR TITLE
ARROW-2262: [Python] Support slicing on pyarrow.ChunkedArray

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

**Comment by xhochy** [comment]
Build fails with 

```
C:\projects\arrow\python\build\temp.win-amd64-3.6\Release\lib.cxx(45584): warning C4244: 'argument': conversion from 'int64_t' to 'int', possible loss of data [C:\projects\arrow\python\build\temp.win-amd64-3.6\Release\lib.vcxproj]
3701C:\projects\arrow\python\build\temp.win-amd64-3.6\Release\lib.cxx(45669): warning C4244: 'argument': conversion from 'int64_t' to 'int', possible loss of data [C:\projects\arrow\python\build\temp.win-amd64-3.6\Release\lib.vcxproj]
```

A review that points me to the problematic code would be appreciated.

File: python/pyarrow/table.pxi [hunk 1]
```diff
@@ -77,6 +77,49 @@ cdef class ChunkedArray:
         self._check_nullptr()
         return self.chunked_array.null_count()
 
+    def __getitem__(self, key):
+        cdef int64_t item, i
+        self._check_nullptr()
+        if isinstance(key, slice):
+            return _normalize_slice(self, key)
+        else:
+            item = int(key)
```
Comment (pitrou) [inline python/pyarrow/table.pxi]: You don't want to allow non-integer types such as float here.

File: python/pyarrow/table.pxi [hunk 2]
```diff
@@ -77,6 +77,49 @@ cdef class ChunkedArray:
...
+            if item >= self.chunked_array.length() or item < 0:
+                return IndexError("ChunkedArray selection out of bounds")
+            for i in range(self.num_chunks):
```
Comment (pitrou) [inline python/pyarrow/table.pxi]: I'm curious, can't this be implemented on the C++ side instead? I suppose C++ users may benefit from such an API as well.

**Comment by pitrou** [comment]
@xhochy what does lib.cxx say around the lines mentioned above? That's assuming your local cython instance produces exactly the same output as AppVeyor does...

**Comment by pitrou** [comment]
@xhochy actually, it's probably because `arrow::ChunkedArray::chunk()` takes a C int but you are trying to pass a int64_t.

File: python/pyarrow/table.pxi (hunk 2 above)
Comment (xhochy) [reply python/pyarrow/table.pxi]: Yes they would benefit from this API but it is much more complicated to implement there as `arrow::ChunkedArray` has no type.

File: python/pyarrow/table.pxi (hunk 1 above)
Comment (xhochy) [reply python/pyarrow/table.pxi]: Changed it

File: python/pyarrow/table.pxi [hunk 3]
```diff
@@ -77,6 +77,52 @@ cdef class ChunkedArray:
...
+    def __getitem__(self, key):
+        cdef int64_t item
+        cdef int i
+        self._check_nullptr()
+        if isinstance(key, slice):
+            return _normalize_slice(self, key)
+        elif isinstance(key, six.integer_types):
+            item = key
+            if item >= self.chunked_array.length() or item < 0:
+                return IndexError("ChunkedArray selection out of bounds")
```
Comment (pitrou) [inline python/pyarrow/table.pxi]: If we allow negative slice bounds, I would expect us to also allow negative indices. Seems like it's time for a `_normalize_index` function?

Comment (wesm) [reply python/pyarrow/table.pxi]: Agreed, perhaps let's handle this as a follow up patch

**Review by wesm — APPROVED** [review]
+1, thanks @xhochy!

**Pull request closed without merging**